async_upnp_client/search.py

```diff
@@ -5,53 +5,51 @@
+from typing import Awaitable, Callable, Mapping, MutableMapping, Optional, Union
 
-from async_upnp_client.ssdp import SSDP_MX
+from async_upnp_client.ssdp import SSDP_IP_V4, SSDP_MX, SSDP_TARGET_V4, SSDP_TARGET_V6
 from async_upnp_client.ssdp import SSDP_ST_ALL
-from async_upnp_client.ssdp import SSDP_TARGET
-from async_upnp_client.ssdp import SSDP_PORT
 from async_upnp_client.ssdp import SsdpProtocol
 from async_upnp_client.ssdp import build_ssdp_search_packet
+from async_upnp_client.ssdp import get_ssdp_socket, get_source_ip_from_target_ip
 
 _LOGGER = logging.getLogger(__name__)
 
+IPvXAddress = Union[IPv4Address, IPv6Address]
+
 
 async def async_search(async_callback: Callable[[Mapping[str, str]], Awaitable],
                        timeout: int = SSDP_MX,
                        service_type: str = SSDP_ST_ALL,
-                       source_ip: Optional[IPv4Address] = None,
+                       source_ip: Optional[IPvXAddress] = None,
                        loop: Optional[AbstractEventLoop] = None,
-                       target_ip: Optional[IPv4Address] = None) -> None:
+                       target_ip: Optional[IPvXAddress] = None) -> None:
     """Discover devices via SSDP."""
-    # pylint: disable=too-many-arguments
-    source_ip = source_ip or IPv4Address('0.0.0.0')
+    # pylint: disable=too-many-arguments,too-many-locals
     loop_: AbstractEventLoop = loop or asyncio.get_event_loop()
 
-    if target_ip:
-        target = (str(target_ip), SSDP_PORT)
-        # We use the standard target in the data of the announce since
-        # many implementations will ignore the request otherwise
-        target_data = SSDP_TARGET
-    else:
-        target = SSDP_TARGET
-        target_data = SSDP_TARGET
+    if target_ip is None:
+        target_ip = IPv4Address(SSDP_IP_V4)
+    if source_ip is None:
+        source_ip = get_source_ip_from_target_ip(target_ip)
 
-    # Create socket.
-    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
-    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-    sock.bind((source_ip.compressed, 0))
+    sock, source, target = get_ssdp_socket(source_ip, target_ip)
+    if target[0] == socket.AF_INET6:
+        ssdp_target = SSDP_TARGET_V6
+    else:
+        ssdp_target = SSDP_TARGET_V4
+    sock.bind(source[4])
 
     async def on_connect(transport: DatagramTransport) -> None:
         """Handle connection made."""
-        packet = build_ssdp_search_packet(target_data, timeout, service_type)
-        transport.sendto(packet, target)
+        packet = build_ssdp_search_packet(ssdp_target, timeout, service_type)
+        transport.sendto(packet, target[4])
 
     async def on_data(_: str, headers: MutableMapping[str, str]) -> None:
         """Handle data."""
         headers['_source'] = 'search'
-        if target_ip:
+        if not target_ip.is_multicast:
             if headers['_address'].partition(":")[0] != f"{str(target_ip)}":
```

Comment:
There is a bug here for ipv6. This won't parse properly for ipv6. I'm not sure if it would be okey to change _address to be separate out port and host. Also the callback currently contain numeric numbers for scope, while the target_ip can contain ascii names for interfaces.